### PR TITLE
issue 14564 - [REG2.067] dmd -property -unittest combination causes compiler error

### PR DESCRIPTION
**Pull request by aG0aep6G**
fixes issue 14564 - [REG2.067] dmd -property -unittest combination causes
compiler error

https://issues.dlang.org/show_bug.cgi?id=14564

The problem is that phobos isn't compatible with `-property`. And it looks like `-property`'s current behaviour isn't desired anymore.

The plan (also in `mars.c`):
- Step 1 (2.068): Ignore -property and warn about it.
- Step 2: Deprecate -property.
- Step 3: Remove -property. Throw an error when it's set.
- Step 4: Possibly reintroduce -property with different semantics.

To do if this gets agreed upon:
- update http://dlang.org/dmd-windows.html and friends,
- changelog entry.

An alternative way to fix issue 14564 would be to make phobos compatible with the current `-property`.

### CODE REVIEW

File: test/fail_compilation/diag8629.d [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
-auto gunc(S s, int n) { return 13; }
 @property grop(S s, int n) { return 14; }
 
 void main()
 {
     S s;
-    assert((s.gunc = 1) == 13);     // line 9
-    assert((s.grop = 1) == 14);     // line 10
+    undef(); // line 8
+    assert((s.grop = 1) == 14);
```
Comment (MartinNowak) [inline test/fail_compilation/diag8629.d]: Just remove the test.

**Comment by MartinNowak** [comment]
@WalterBright please approve.

File: src/mars.c [hunk 2]
```diff
@@ -802,8 +801,29 @@ Language changes listed by -transition=id:\n\
...
+            {
+                /*NOTE: -property used to disallow calling non-properties
+                without parentheses. This behaviour has fallen from grace.
+                Phobos dropped support for it while dmd still recognized it, so
+                that the switch has effectively not been supported. Time to
+                remove it from dmd.
+                Step 1 (2.068): Ignore -property and warn about it. */
+                warning(Loc(), "The -property switch has no effect since "
+                    "version 2.068.0. It's going to be deprecated and "
+                    "eventually removed in future versions.");
```
Comment (MartinNowak) [inline src/mars.c]: You have to warn after the whole argument parsing is done b/c they way it's done now `-w -property` would warn while `-property -w` wouldn't, i.e. `global.params.warning` might still be false.

**Comment by MartinNowak** [comment]
From Walter:
> I'm a bit reluctant to pull it for 2.068 out of fear of it being disruptive this late in the beta cycle. 

I agree with him, it's a bit too disruptive to add this to a release candidate, e.g. everyone using `-w -property` will have to adjust their build scripts.
Please reopen this to target master so we can go on with this for the next release.

**Comment by aG0aep6G** [comment]
> Please reopen this to target master so we can go on with this for the next release.

https://github.com/D-Programming-Language/dmd/pull/4862

**Pull request closed without merging**